### PR TITLE
WIP: topic creation/deletion behaviour.

### PR DESCRIPTION
**Pull request by martinbremmer**
Originally, this was a PR to improve the create domain feature. It was noticed that the topic creation/deletion behaviour changed compared to master. A quick change was suggested to solve that. However, that wasn't accepted. Currently, the master contains the revised behaviour (see #319).

This PR is changed to a WIP to not lose the history of the topic creation/deletion discussion nor lose the proposed (but declined) solution.

Original PR description:
> See issue #300.
> 
> This upgrade impacts the entity deletion quite a bit, as indicated in the WIP pull-request #302.
> 
> That #302 PR contains a WIP commit regarding dds_create_domain() and entity deletions.
> 
> This PR takes that commit, changed some function headers, added some tests and refactored the WIP commit a bit.
> 
> The main difference after the refactor is that only ```pinning for deletion``` determines if the ```closing``` flag should be set. It's not done in the ```reduce refcount``` functions anymore.
> By changing ```pinning for deletion``` a little bit more, the 'normal' entity delete sequence can be followed for (builtin) topics, domains and cyclonedds top entity (no need for specific repins).

### CODE REVIEW

File: src/core/ddsc/src/dds_entity.c [hunk 1]
```diff
@@ -80,11 +80,34 @@ const ddsrt_avl_treedef_t dds_entity_children_td = DDSRT_AVL_TREEDEF_INITIALIZER
...
 void dds_entity_add_ref_locked (dds_entity *e)
 {
   dds_handle_add_ref (&e->m_hdllink);
 }
 
+void dds_entity_drop_ref (dds_entity *e)
+{
+  if (dds_handle_drop_ref (&e->m_hdllink))
+  {
+    dds_return_t ret = dds_delete_impl(e->m_hdllink.hdl, DIS_EXPLICIT);
```
Comment (eboasson) [inline src/core/ddsc/src/dds_entity.c]: There was (some) method behind the madness of my "WiP" commit in setting the ``CLOSING`` flag on the handle in ``dds_handle_drop_ref`` (and ``dds_handle_unpin_and_drop_ref``). I'm no way trying to defend the body of this function (and the next) in that commit, that was simply an ugly hack along the way.

The way you do it here looks perfectly fine, but what it does allow is for some thread to increment the refcount after it dropped to 0 but before ``dds_delete_impl`` is called, if there is a path that allows another thread to do so. One has to wonder whether that's an actual problem, and there are two perspectives on it:

1. Is there any way in which that can happen? — If the answer is "no", then all is well.
2. If it happens, does it behave in an acceptable manner? — If it always behaves in an acceptable manner (not crashing for example) then all is well, too.

Re 1: for ``dds_entity_drop_ref``, it really is only a topic that can be passed in, and it would already have to be in the "deferred delete" state (otherwise the implied reference from the handle will keep it alive). I'm pretty sure it is possible, but it can be prevented by checking in ``dds_create_reader`` and ``dds_create_writer``. That is a manageable case.

``dds_entity_unpin_and_drop_ref`` is only used for the global "cyclone" entity and the domain entities. Those are also really special cases, sneak around the regular destruction path and have additional locking to prevent such games. I just haven't figured that one out yet ...

Re 2: For topics, it could cause the call to ``delete`` to return ``ALREADY_DELETED``, but I think nothing worse than that. One could change the assert, one could even decide that ``ALREADY_DELETED`` doesn't make sense and that it should always return ``OK`` anyway.

For the domain and cyclonedds objects, I think it is equivalent to simply calling ``dds_delete`` from the application code. It may therefore result in deleting a domain/participant that was being created in parallel, but I do think that afterward that new one either exists and functions normally, or doesn't exist at all. Given that it is somewhat daft use of the API, I'd say that is acceptable.

In short, your simplification looks reasonable; in this form it will work better than the code currently in master (e.g., calling ``dds_delete`` twice on a topic is deadly now); and the possibly problematic cases seem to be rather far out there in terms of realistic use.

It'd be interesting to try testing it a bit, for example by creating and deleting participants in some domain on multiple threads.

Comment (martinbremmer) [reply src/core/ddsc/src/dds_entity.c]: I've added a torture test that showed that a wrong mutex was used when adding a participant to the domain as a child.

**Comment by martinbremmer** [comment]
There's some behaviour change regarding creation/deletion of topics. I'm still looking into that.

The current solution direction will be to 'clone' a topic (creating a new handle) when a duplicate topic is created on a participant.

File: src/core/ddsc/src/dds_entity.c (hunk 1 above)
Comment (martinbremmer) [reply src/core/ddsc/src/dds_entity.c]: It still crashes intermittently when compiling with gcc.
Somehow, I can't reproduce it with clang.

Comment (martinbremmer) [reply src/core/ddsc/src/dds_entity.c]: There are two torture tests. One triggers an exiting issue (the crash) that is somewhat outside the scope of this PR and is related to actually init/deinit the library. For that reason, I've disabled that torture test.

File: src/core/ddsc/src/dds_participant.c [hunk 2]
```diff
@@ -126,9 +126,9 @@ dds_entity_t dds_create_participant (const dds_domainid_t domain, const dds_qos_
   pp->m_builtin_subscriber = 0;
 
   /* Add participant to extent */
-  ddsrt_mutex_lock (&dds_global.m_entity.m_mutex);
+  ddsrt_mutex_lock (&dom->m_entity.m_mutex);
```
Comment (eboasson) [inline src/core/ddsc/src/dds_participant.c]: 😳

Thanks for catching that one!

**Comment by martinbremmer** [comment]
I don't have strong feelings either way. It's just that the 'clone' option currently implemented is more intuitive (at least to me) when considering deletions.

It also makes my live a lot easier regarding C++. But as you said, that should be solvable.

**Comment by eboasson** [comment]
I have merged #319 which is basically this one minus the last commit. That commit is interesting, so don't throw that away just yet please, @martinbremmer. (In fact, you could turn it into a WIP PR if you wish.)
